### ISSUE TITLE
dynamic visualization of nodes and channels in Oak Runtime

### ISSUE THREAD

**Issue by tiziano88** [project-oak/oak#913]
#902 added support to the Oak Runtime for an HTTP endpoint that displays the current state of the nodes and channels in a visual way (built on [d3.js](https://d3js.org/) but displaying a [DOT](https://www.graphviz.org/doc/info/lang.html) formatted file). It would be nice to make this a more native and dynamic visualization by encoding the state as a JSON object, and using some fancy d3 visualization.

@juliettepretot I hear you like Javascript, what do you think about giving this a try? I know it's quite vague, feel free to ask questions or suggest alternative approaches!

**Comment by tiziano88** [comment]
Perhaps something along these lines: https://observablehq.com/@d3/mobile-patent-suits

**Comment by juliettepretot** [comment]
That's a lovely idea, I'll look into it. :)

**Comment by juliettepretot** [comment]
Having looked into this more, here's what my game plan looks like: 

My first goal would to make the visualizer dynamically update. 

Technically I would implement this with [Server-sent events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events) (that's a uni-directional stream of events from the server to the client). Initially this stream would merely send the updated rendered Graphviz Dot graph to browser, when nodes are created/deleted.

On the (rust) server-side, I was thinking about [warp](https://github.com/seanmonstar/warp) for this. It's a web server framework built on top of hyper, and implements [Server-sent events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events) and [is recommended as higher-level abstraction by hyper](https://github.com/hyperium/hyper#low-level).

My initial PR would be refactor the existing introspection server logic to use warp.

Later I would add dynamic updating with the event stream. I will probably need some help writing the async rust code that triggers the server side event when the runtime's `node_infos` hashmap is changed. 

Do people think that's a fair approach? Does the goal makes sense, and does my plan of getting there appear sound? cc @daviddrysdale @tiziano88 

On the client side it would be pretty standard, I'd use the existing [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) class to consume the event stream, and use the built in [d3 Graphviz primitives](https://github.com/magjac/d3-graphviz#creating-transitions) to animate between the updated graphs. 

---

In my mind this adds initial value, and the infrastructure for it can be used to implement my two stretch goals: Implementing timescrubbing (go back in time), and adding data-flow to the visualizer.

Those can reuse the event stream. At that point it probably makes sense to move away from sending pre-rendered graphs from the introspection server, and instead serialize the node_infos hashmap to json and leave it up to the client to render that data in a helpful manner.

**Comment by tiziano88** [comment]
Thanks for the summary @juliettepretot , very promising!

My main concern with this plan is that it requires complex logic to be added to the Oak Runtime, which increases the TCB, and also the cognitive overhead for current and future maintainers.

Have you considered a simpler model in which the Runtime keeps track of the history internally (i.e. has timestamps attached to nodes and messages, etc.), but the client just periodically polls the endpoint and gets the entire list of events as JSON? Or do you think it would become too large too quickly?

Even if you think that's not ideal as an end goal, I think it would be worth having this as an explicit intermediate step.

A couple other things to think about:

- event sources vs websockets (or are they the same thing?)
- can this mechanism be easily disabled for "production" builds of the Runtime?

**Comment by juliettepretot** [comment]
@tiziano88 

Thanks for chiming in! I considered the polling approach as well. It does reduce the complexity around needing to push updates.

Since I'm new to the project it's tricky for me to estimate how complex keeping track of of the history in the runtime is. With nodes being created/deleted, temporary channels (eg the invocation pattern), and messages moving through them I can imagine it getting complex. But I'd trust your judgement on its feasibility of it over my impression. 

I had some hesitation to go with that route since keeping internal history seemed to require a lot of runtime changes, going to your point around TCB size.  

If I were to go down this route, it seems like I'd be three steps

1. Implementing a client side mechanism to poll for updates & compare. (Could be a simple `<meta http-equiv="refresh" content="5">`, but I can make it nicer with client side logic)
2. Serializing the node_infos hashmap as JSON and moving rendering logic to the client. (Might need help with serializing it)
3. Keeping track of history in the runtime (I most definitely need help with that).

---

To answer your other two questions: 

- Event sources / web sockets are distinct. Event sources are uni-directional, web sockets are bi-directional. Since we it seems we don't need bi-directionality, the simpler event sources seemed like the way to go. 
- I don't the something that would block this disabling, but I will caution that I'm new to the project, may be overlooking concerns.

**Comment by juliettepretot** [comment]
On a higher level it seems that an argument for moving towards sending over a JSON file, as opposed to the current prerendered graph `node_infos` would be reduction of the TCB? Afaik all the rust code currently powering the introspection server counts towards the TCB?

**Comment by daviddrysdale** [comment]
I wonder if we can get the best of both/all worlds with some separation of interface from implementation.

At the "interface" level, we could add an internal macro that indicates that an event changing internal data structures has happened.  This macro could be easily compiled out in non-`oak_debug` builds, and would only be needed in a dozen or so places in the Runtime, thus reducing the concern about Runtime complexity and TCB size.  Roughly something like

 - `event!("node", "create", node_id)`
 - `event!("node", "destroy", node_id)`
 - `event!("channel", "create", node_id, channel_id)`
 - `event!("channel", "destroy", node_id, channel_id)`
 - `event!("handle", "create", node_id, handle, channel_id)`
 - `event!("handle", "destroy", node_id, handle, channel_id)`
 - `event!("message", "enqueue", node_id, channel_id, included_handles)`
 - `event!("message", "dequeue", node_id, channel_id, acquired_handles)`

with `"destroy"` `event!`s triggered from `Drop` implementations where possible (for robustness), and no `async` code needed in the main Runtime code.

This gives a stream of events where the information contained in them should<sup>*</sup> be replayable to give a complete picture of state of the internal data structures (modulo inter-thread reorderings, but I *think* there shouldn't be non-transient observable differences).

The implementation of the `event!` macro could then either have a single implementation that immediately consumes it, or could pubsub out to multiple consumers.  (Which is a decision that can be revisited later without needing to change any code in the main Runtime.)

Possible consumers of this stream would be:
 - A simple tracker that keeps a current graph of data structures, ready to be retrieved by something like `runtime.graph()` as a Dot graph.
 - A simple counter that counts the number of extant nodes and channels, ready to be retrieved by something like `runtime.object_count()` 
 - A web server that passes on `event!`s as server-sent events to its clients.
 - A history dumper that stores the event stream to file, adding timestamps, allowing offline playback later.

All of these should/could be interchangeable, independent of each other, and (in particular) independent of the core Runtime &ndash; if we get the initial set of `event!` information right, adding any of these consumers should result in zero changes to the core Runtime code.

<hr/>

There is one worry with this approach though: the "*should*" marked [*] above.  This all assumes that a complete stream of changes to Runtime data structures exactly matches the current actual state of Runtime data structures.  If there are bugs in the Runtime's internal tracking of data structures, then this may not be true &ndash; and then our introspection facility will be lying to us and making debugging much more difficult.  This isn't entirely a theoretical concern, either &ndash; one of the main reasons I wrote the introspection stuff in the first place was to help find exactly these kinds of internal bugs in the Runtime.  So depending on how confident we feel, we might want to keep the existing recursive `runtime.graph()` implementation as a cross-check for a while.

**Comment by daviddrysdale** [comment]
BTW, the sketch above is a bit [stringly-typed](https://wiki.c2.com/?StringlyTyped); in Rust-land I'd guess we'd end up with something stronger (or maybe even a protobuf definition for events)&hellip;

**Comment by juliettepretot** [comment]
Had a quick conversation with @tiziano88 & @daviddrysdale, we agreed to go with the approach of events, to decouple interface from implementation. `

### Getting out of sync
As @daviddrysdale mentioned, unlike the runtime keeping track of history internally, this approach does run the risk of "getting out of sync" if events aren't fired. But since events only need to be fired in a few places, we're ok with this for now. We will consider approaches to mitigate this later if/when they become needed.  

### Target
My initial target would be building the logic to send these events, and a simple consumer that would consume events and write them to the disk. 

Once this is done, I'll add logic that enables the web client to use the history of past events to implement a time-scrubbing time UX. Dynamic updating can be implemented later either via polling or via a event consumer that's a web-server streaming received events to the web client. 


### Next steps for me
1. Learn about how to send events out of the runtime & to consumers. @daviddrysdale mentioned I can use (non-Oak) channels for that. I'm new to this, but that sound like an ABI? My intuition would be that protobufs would be used for sending that data from the runtime to consumers. Some more specific guidance here would be amazing :) 
2. Implement the logic for sending events in the Rust runtime
3. Write the basic consumer that writes received events to the disk. 

cc @daviddrysdale @tiziano88, does that sound good?

**Comment by tiziano88** [comment]
The channels @daviddrysdale and I were referring to are either https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html or https://docs.rs/tokio/0.2.21/tokio/sync/mpsc/fn.channel.html , and only available within the Oak Runtime itself, so they don't require an ABI at all.

I would suggest to start defining a schema (perhaps in protobuf form) to describe the kind of events we want to be able to capture (I expect this is going to look like a `message Event` with a number of metadata fields and a single `oneof` field for the event details.

Once that is in place, then add methods to the runtime object to generate such events, and a no-op implementation of a handler for such events.

Then, make this handler just push events over one of the channels mentioned earlier, and pass the receiving end of the channel to the introspection aux server, which will decide what to do with it.

For the production version of the `oak_loader`, the event handler will just have a no-op implementation, which means the compilers should effectively optimize away all the calls to the methods used to generate the events (since it should detect that such events are never used).

**Comment by juliettepretot** [comment]
That sounds good, thank you for the additional details, will get to work on that. :)

**Comment by juliettepretot** [comment]
I ran into a bit of a roadblock with implementing the channel dropped event, since it's not possible to do so from within a `Runtime` method. 

It seems we need to either rethink our approach by converting the event method to a func (with some potential other implications), or potentially it could make sense to omit that event altogether. 

See https://project-oak.slack.com/archives/CHE9E13C3/p1594995744232500 for details. 

My plan at the moment is to proceed with implementing the other events, as well as the aux server, and discuss this further with @daviddrysdale or @tiziano88 when they're back from vacation. :) 🏝

**Comment by juliettepretot** [comment]
@tiziano88 I'm currently implementing the channel to send events to a aux server, and it turns out to be a bit tricker than I anticipated. The reason being that both rust's and tokio's channels you mentioned only support a single receiver.

To me it seems that means I need to a set of shenanigans to move over that end over to the aux server, potentially by storing the receiving end in its own Arc & Mutex. And there could only ever be one consumer. 

https://docs.rs/tokio/0.2.21/tokio/sync/broadcast/index.html Looks easier, do you think I can use that instead?

**Comment by tiziano88** [comment]
Thanks @juliettepretot , I think conceptually using a broadcast channel makes sense here, since the intention is really for the runtime to broadcast events to any aux server that cares about them. The downside is that, AFAICT, a broadcast channel would hold onto every message until all outstanding readers have consumed it; since most aux servers would never consume such events, it seems that it would lead to the message queue growing forever by accumulation of new objects?

**Comment by juliettepretot** [comment]
Agreed that conceptually that makes more sense. :) 

I don't think the "holding onto messages for each consumer" would become a problem. In order to become a consumer, the aux server explicitly needs to call the `subscribe` method on the receiver. Only our introspection server would do that. :)

**Comment by juliettepretot** [comment]
@tiziano88 Agreed that conceptually that makes more sense. :) 

I don't think the "holding onto messages for each consumer" would become a problem. In order to become a consumer, the aux server explicitly needs to call the `subscribe` method on the receiver. Only our introspection server would do that. :)

**Comment by juliettepretot** [comment]
Going back to our decision to use a `Weak` to reference the runtime within the `Channel` struct, I might need some direction with lifetime annotations. They're still a novel concept that I'm wrapping my head around. 

I understand our intention, which is to (weakly) reference the Runtime from the Channel. Based off how our program works, a Channel should never outlive the Runtime, hence the reference held in the `Weak` will always be present. 

However I'm struggling to communicate this to the Rust compiler, which requires lifetime annotations for all references held in a struct. 

Eg I would need one right here:

```rs
pub struct Channel {
   ...

    /// Weak reference to the Runtime used for sending introspection events.
    /// Requires life time annotation for `&Runtime`
    runtime: Weak<&Runtime>,
}
```

Essentially it seems the compiler would like to know the lifetime of `Runtime`. Whereas our intention is: Assume it'll be a valid reference for the lifetime of `Channel`.

If I go through with adding the annotations I basically end up passing lifetime parameters all over the runtime, which doesn't seem like what we want. 

It seems that perhaps we should use the `'static` lifetime? In that case however I'm not sure how to pass the non-static `self` reference from within the runtime to the `Channel::new` which then expects a Runtime with static lifespan. 

cc @tiziano88 @daviddrysdale

**Comment by tiziano88** [comment]
*I think* you just need an `Arc<Runtime>` (no reference in the type), since the liveness check is effectively done at runtime by calling [`upgrade`](https://doc.rust-lang.org/beta/std/sync/struct.Weak.html#method.upgrade) on it. Or have you tried that and didn't work?

**Comment by juliettepretot** [comment]
@tiziano88 It appears the the [downgrade](https://doc.rust-lang.org/beta/std/sync/struct.Arc.html#method.downgrade) method requires a reference to an Arc, hence how I end up in the trenches dealing with references. Is the something I'm over looking?

**Comment by tiziano88** [comment]
Sorry, my mistake, I meant that you need a `Weak<Runtime>`, which you would get from the `Arc<Runtime>` that is already embedded here:

https://github.com/project-oak/oak/blob/347d3a5d12d13a9bac031d6f443dee5b5a4881c7/oak/server/rust/oak_runtime/src/proxy.rs#L51

**Comment by juliettepretot** [comment]
@tiziano88 ah yup, that makes a lot of sense. I understand now. Thanks! :)

The reference came from creating an Arc from within a Runtime method which ofc only contains a reference to itself, hence the `Arc` and subsequently `Weak` contained a reference.

I had mistakenly thought the reference came from [downgrade](https://doc.rust-lang.org/beta/std/sync/struct.Arc.html#method.downgrade) requiring a reference to an `Arc`, but that's something else. :))

**Comment by juliettepretot** [comment]
On an adjacent topic, I would appreciate getting your thoughts reconstructing past applications states for time scrubbing. 

On a high level, I think there's two approaches here: A) do it on the web client, B) do it in the Rust introspection server. 

A) would mean serializing the `introspection_event_queue` to JSON & sending it over to the client. The client would use JavaScript to generate an application graph at any given point in time based off it. 

Pros: 
- State & UI is handled entirely client side
- No requests to the aux server needed when scrubbing
- ⚡️ likely faster for me to write, since I'm much more fluent in JS.

Cons: 
- Potentially costly serialization of events to JSON
- Needing to rewrite our logic for generate a graph in JS. 
- Less Rust, more JavaScript. People like Rust more than JS. 


B) would mean all logic for computing past applications states is done in our Rust aux server. To display past states, the web based introspection UI would send a request to the introspection server, which then generate an application graph for it. 

Pros: 
- Builds on top of our existing Rust code to generating application graphs
- More Rust, less JavaScript. People like Rust more than JS. 

Cons: 
- More APIs between the aux server & client side code are needed
- Interlinked state between the introspection client(s) and aux server could lead to edge-cases, bugs etc. (Eg `introspection_event_queue` drops older events, but an old client still shows scrubber to go further into the past.)


To me option A) seems most sensible, but what do you all think? cc @daviddrysdale @tiziano88

**Comment by tiziano88** [comment]
I strongly prefer approach A too.

B seems mostly an optimization for cases in which A does not scale well, but I would postpone until it actually becomes a substantial problem. In the meanwhile we can learn a lot about the expected use cases.

**Comment by juliettepretot** [comment]
SGTM. I would add that imho there's solutions to the scaling issues that might arise with A that wouldn't necessitate switching over to B, though we still might choose to later do that for different reasons.

**Comment by juliettepretot** [comment]
I've new question on approach that I'd be intrigued to get your opinions on. :) 

So now that we have a list of events, we can deduct application state at any given point in time, and then use that application state to render UI visualising that point in time. 

My question is which data structure to use for the deducted application state. 

a) Use our own structs on the client, which are simplified but analogous to the structs used in the Runtime. eg seperate versions of `Node`, `Channel`, `ChannelHalf`. (This could be either JS or Rust client code)

b) Literally reuse the structs from runtime. (This needs to be Rust client code)

cc @tiziano88 @daviddrysdale

**Comment by tiziano88** [comment]
I would imagine that the client code defines its own structures, driven by UI requirements; an intermediate layer (in our case I guess in the introspection server code itself) converts from the internal structs to the UI structs, and also serializes them to JSON (or protobuf). So I guess this sounds like your option a?

**Comment by juliettepretot** [comment]
Hmm maybe. 

I don't think there actually are any UI requirements that define the structs on the client. Defacto they'd be super similar to how state is defined in the runtime. The advantage to b) would be that the code is quite closely interlinked via, which could prevent code rotting, but also might add complexity. 

To give a summary of the architecture I have in mind in case it's not entirely clear:

Runtime logs events -> introspection server serializes them and provides exposes them via an API -> a browser client loads the list of events, deserializes them, constructs an application state from them, and then paints UI based of the application state. 

Does that make sense? :) 

@tiziano88
